specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,1028 @@
+functions that can be called by default in the entry block are `qis` or `rt`
+functions defined in the instruction set and profile. This restriction is removed
+if a backend opts into **Bullet 10** which allows for user-defined functions.
+
+## Quantum Instruction Set
+
+For a quantum instruction set to be fully compatible with the adaptive profile, it
+must satisfy the following three requirements:
+
+- All functions must return `void`, `iN`, or `fN` types (`i1` and `i64` for example)
+  and can only take in `%Qubit*`, `%Result*`, `iN`, or `fN` types as arguments.
+  Additionally, arguments to the quantum instruction set cannot have dynamically
+  computed arguments by default. They can however refer to static global values
+  that are defined by linking (for example
+  `call @__quantum__qis__rz__body(double @angle, %Qubit* null)` where `@double`
+  is a static global who gets a new rotation angle linked at each shot).
+  Dynamically computed arguments can be supplied to instruction set functions
+  when adaptive profile programs use a classical data type from **Bullet 9**.
+  Functions that measure qubits must take the qubit pointer(s) as well as the
+  result pointer(s) as arguments.
+
+- Functions that perform a measurement of one or more qubit(s) must be marked
+  with a custom function attribute named `irreversible`. The use of
+  [attributes](#attributes) in general is outlined in the corresponding section.
+
+For more information about the relationship between a profile specification and the
+quantum instruction set, we refer to the paragraph on [Bullet 1](#bullet-1-quantum-transformations)
+in the introduction of this document. For more information about how and when
+the QIS is resolved, as well as recommendations for front- and backend
+developers, we refer to the document on [compilation stages and
+targeting](../Compilation_And_Targeting.md).
+
+## Classical Instructions
+
+The following instructions are the LLVM instructions that were permitted
+within a base profile-compliant program:
+
+| LLVM Instruction         | Context and Purpose                                                                               | Rules for Usage                                                                                             |
+| :----------------------- | :------------------------------------------------------------------------------------------------ | :---------------------------------------------------------------------------------------------------------- |
+| `call`                   | Used within a basic block to invoke any one of the declared QIS functions and run-time functions. | May optionally be preceded by a [`tail` marker](https://llvm.org/docs/LangRef.html#call-instruction).       |
+| `br`                     | Used to branch from one block to another in the entry point function.                             | The branching must be unconditional and occurs as the final instruction of a block to jump to the next one. |
```

Comment:
maybe we should just list the classical computations on i1 that we require even without optional capabilities here